src/Data/Tree/AVL/Iteration.hs

```diff
@@ -1,210 +1,75 @@
-{-# LANGUAGE FlexibleInstances     #-}
-{-# LANGUAGE MultiParamTypeClasses #-}
-{-# LANGUAGE TypeSynonymInstances  #-}
 
-module Data.Tree.AVL.Iteration where
+module Data.Tree.AVL.Iteration
+    ( walkDFS
+    , foldIf
+    , fold
+    )
+  where
+
+import Control.Lens ((^.))
+import Control.Monad.Writer.Strict (WriterT(runWriterT), tell, lift)
+
+import qualified Data.Set as Set
+
+import Data.Tree.AVL.Internal
+
+-- | Traverses tree in DFS manner meeting keys in acsending order.
+walkDFS
+  :: forall h k v m b res
+  .  Retrieves h k v m
+  =>  ( b
+      , MapLayer h k v (Maybe h) -> b -> b
```

Comment:
What's the point of accepting the tuple instead of simply having four arguments? Same applies to other two functions in this module.